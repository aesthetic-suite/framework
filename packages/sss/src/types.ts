import {
  CSST,
  FontFace as BaseFontFace,
  Keyframes as BaseKeyframes,
  Declarations,
  Property,
  Value,
  Variables,
} from '@aesthetic/types';
import Block from './Block';

// PROPERTIES

export interface AnimationProperty {
  delay?: CSST.Property.AnimationDelay;
  direction?: CSST.Property.AnimationDirection;
  duration?: CSST.Property.AnimationDuration;
  fillMode?: CSST.Property.AnimationFillMode;
  iterationCount?: CSST.Property.AnimationIterationCount;
  name?: CSST.Property.AnimationName;
  playState?: CSST.Property.AnimationPlayState;
  timingFunction?: CSST.Property.AnimationTimingFunction;
}

export interface BackgroundProperty {
  attachment?: CSST.Property.BackgroundAttachment;
  clip?: CSST.Property.BackgroundClip;
  color?: CSST.Property.BackgroundColor;
  image?: CSST.Property.BackgroundImage;
  origin?: CSST.Property.BackgroundOrigin;
  position?: CSST.Property.BackgroundPosition<Value>;
  repeat?: CSST.Property.BackgroundRepeat;
  size?: CSST.Property.BackgroundSize<Value>;
}

export interface BorderProperty {
  color?: CSST.Property.BorderColor;
  style?: CSST.Property.BorderStyle;
  width?: CSST.Property.BorderWidth<Value>;
}

export interface ColumnRuleProperty {
  color?: CSST.Property.ColumnRuleColor;
  style?: CSST.Property.ColumnRuleStyle;
  width?: CSST.Property.ColumnRuleWidth<Value>;
}

export interface FlexProperty {
  basis?: CSST.Property.FlexBasis<Value>;
  grow?: CSST.Property.FlexGrow;
  shrink?: CSST.Property.FlexShrink;
}

export interface FontProperty {
  family?: CSST.Property.FontFamily;
  lineHeight?: CSST.Property.LineHeight<Value>;
  size?: CSST.Property.FontSize<Value>;
  stretch?: CSST.Property.FontStretch;
  style?: CSST.Property.FontStyle;
  system?: string;
  variant?: CSST.Property.FontVariant;
  weight?: CSST.Property.FontWeight;
}

export interface ListStyleProperty {
  image?: CSST.Property.ListStyleImage;
  position?: CSST.Property.ListStylePosition;
  type?: CSST.Property.ListStyleType;
}

export interface MarginProperty {
  bottom?: CSST.Property.MarginBottom<Value>;
  left?: CSST.Property.MarginLeft<Value>;
  leftRight?: CSST.Property.MarginLeft<Value> | CSST.Property.MarginRight<Value>;
  right?: CSST.Property.MarginRight<Value>;
  top?: CSST.Property.MarginTop<Value>;
  topBottom?: CSST.Property.MarginTop<Value> | CSST.Property.MarginBottom<Value>;
}

export interface OffsetProperty {
  anchor?: CSST.Property.OffsetAnchor<string>;
  distance?: CSST.Property.OffsetDistance<string>;
  path?: CSST.Property.OffsetPath;
  position?: string; // NOT UPSTREAM
  rotate?: CSST.Property.OffsetRotate;
}

export interface OutlineProperty {
  color?: CSST.Property.OutlineColor;
  style?: CSST.Property.OutlineStyle;
  width?: CSST.Property.OutlineWidth<Value>;
}

export interface PaddingProperty {
  bottom?: CSST.Property.PaddingBottom<Value>;
  left?: CSST.Property.PaddingLeft<Value>;
  leftRight?: CSST.Property.PaddingLeft<Value> | CSST.Property.PaddingRight<Value>;
  right?: CSST.Property.PaddingRight<Value>;
  top?: CSST.Property.PaddingTop<Value>;
  topBottom?: CSST.Property.PaddingTop<Value> | CSST.Property.PaddingBottom<Value>;
}

export interface TextDecorationProperty {
  color?: CSST.Property.TextDecorationColor;
  line?: CSST.Property.TextDecorationLine;
  style?: CSST.Property.TextDecorationStyle;
  thickness?: CSST.Property.TextDecorationThickness<Value>;
}

export interface TransitionProperty {
  delay?: CSST.Property.TransitionDelay;
  duration?: CSST.Property.TransitionDuration;
  property?: CSST.Property.TransitionProperty;
  timingFunction?: CSST.Property.TransitionTimingFunction;
}

export type ListableProperty<B, T> = B | T | (B | T)[];

// SYNTAX

export type CompoundPropertyTypes = 'animationName' | 'fontFamily';

export type ExpandedPropertyTypes =
  | 'animation'
  | 'background'
  | 'border'
  | 'borderBottom'
  | 'borderLeft'
  | 'borderRight'
  | 'borderTop'
  | 'columnRule'
  | 'flex'
  | 'font'
  | 'listStyle'
  | 'margin'
  | 'offset'
  | 'outline'
  | 'padding'
  | 'textDecoration'
  | 'transition';

export interface Properties
  extends Omit<CSST.StandardProperties<Value>, CompoundPropertyTypes | ExpandedPropertyTypes> {
  animation?: CSST.Property.Animation | AnimationProperty;
  animationName?: CSST.Property.AnimationName | Keyframes;
  background?: CSST.Property.Background<Value> | BackgroundProperty;
  border?: CSST.Property.Border<Value> | BorderProperty;
  borderBottom?: CSST.Property.BorderBottom<Value> | BorderProperty;
  borderLeft?: CSST.Property.BorderLeft<Value> | BorderProperty;
  borderRight?: CSST.Property.BorderRight<Value> | BorderProperty;
  borderTop?: CSST.Property.BorderTop<Value> | BorderProperty;
  columnRule?: CSST.Property.ColumnRule<Value> | ColumnRuleProperty;
  flex?: CSST.Property.Flex<Value> | FlexProperty;
  font?: CSST.Property.Font | FontProperty;
  fontFamily?: ListableProperty<CSST.Property.FontFamily, FontFace>;
  listStyle?: CSST.Property.ListStyle | ListStyleProperty;
  margin?: CSST.Property.Margin<Value> | MarginProperty;
  offset?: CSST.Property.Offset<Value> | OffsetProperty;
  outline?: CSST.Property.Outline<Value> | OutlineProperty;
  padding?: CSST.Property.Padding<Value> | PaddingProperty;
  textDecoration?: CSST.Property.TextDecoration<Value> | TextDecorationProperty;
  transition?: CSST.Property.Transition | TransitionProperty;
}

export type FallbackProperties = CSST.StandardPropertiesFallback<Value>;

export type Rule = Declarations<Properties>;

export interface RuleMap<T> {
  [selector: string]: T;
}

export interface FontFace extends BaseFontFace {
  local?: string[];
  srcPaths: string[];
}

export interface Import {
  path: string;
  media?: string;
  url?: boolean;
}

export type Keyframes = BaseKeyframes<Rule>;

export type Viewport = CSST.AtRule.Viewport<Value>;

// TODO Add upstream to csstype
export type PageMargins =
  | '@top-left-corner'
  | '@top-left'
  | '@top-center'
  | '@top-right'
  | '@top-right-corner'
  | '@bottom-left-corner'
  | '@bottom-left'
  | '@bottom-center'
  | '@bottom-right'
  | '@bottom-right-corner'
  | '@left-top'
  | '@left-middle'
  | '@left-bottom'
  | '@right-top'
  | '@right-middle'
  | '@right-bottom';

export type PageBlock = Properties & {
  bleed?: string;
  marks?: string;
  size?: string;
} & {
    [K in PageMargins]?: PageBlock;
  };

export type PagePseudos = ':blank' | ':first' | ':left' | ':right';

export type Page = PageBlock & { [K in PagePseudos]?: PageBlock };

// LOCAL STYLE SHEET

export type LocalAtRule =
  | '@fallbacks'
  | '@media'
  | '@selectors'
  | '@supports'
  | '@variables'
  | '@variants';

export type LocalBlock = Rule & {
  '@fallbacks'?: FallbackProperties;
  '@media'?: { [mediaQuery: string]: LocalBlock };
  '@selectors'?: { [selector: string]: LocalBlock };
  '@supports'?: { [featureQuery: string]: LocalBlock };
  '@variables'?: Variables;
  '@variants'?: { [variant: string]: LocalBlock };
};

export type LocalBlockNeverize<T> = {
  [K in keyof T]: K extends keyof LocalBlock ? T[K] : never;
};

export type LocalStyleSheet = RuleMap<string | LocalBlock>;

export type LocalStyleSheetNeverize<T> = {
  [K in keyof T]: T[K] extends string ? string : LocalBlockNeverize<T[K]>;
};

// GLOBAL STYLE SHEET

export type GlobalAtRule =
  | '@font-face'
  | '@import'
  | '@keyframes'
  | '@page'
  | '@root'
  | '@variables'
  | '@viewport';

export interface GlobalStyleSheet {
  '@font-face'?: { [fontFamily: string]: FontFace | FontFace[] };
  '@import'?: (string | Import)[];
  '@keyframes'?: { [animationName: string]: Keyframes };
  '@page'?: Page;
  '@root'?: LocalBlock;
  '@variables'?: Variables;
  '@viewport'?: Viewport;
}

export type GlobalStyleSheetNeverize<T> = {
  [K in keyof T]: K extends keyof GlobalStyleSheet ? GlobalStyleSheet[K] : never;
};

// MISC

export type AtRule = LocalAtRule | GlobalAtRule;

export interface NestedBlockParams {
  specificity: number;
}

export type OnProcessProperty = (property: Property, value: Value) => void;

export type Processor<T> = (value: T, onProcess: OnProcessProperty) => Value | undefined | void;

export interface ProcessorMap {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [key: string]: Processor<any>;
}

// EVENTS

export type BlockConditionListener<T extends object> = (
  parent: Block<T>,
  condition: string,
  value: Block<T>,
) => void;

export type BlockNestedListener<T extends object> = (
  parent: Block<T>,
  name: string,
  value: Block<T>,
  params: NestedBlockParams,
) => void;

export type BlockPropertyListener<T extends object> = (
  parent: Block<T>,
  name: string,
  value: unknown,
) => void;

export type BlockListener<T extends object> = (block: Block<T>) => void;

export type ClassNameListener = (selector: string, className: string) => void;

export type FontFaceListener<T extends object> = (
  fontFace: Block<T>,
  fontFamily: string,
  srcPaths: string[],
) => void;

export type ImportListener = (path: string) => void;

export type KeyframesListener<T extends object> = (
  keyframes: Block<T>,
  animationName: string | undefined,
) => string;

export type RuleListener<T extends object> = (selector: string, block: Block<T>) => void;

export type VariableListener = (name: string, value: Value) => void;
