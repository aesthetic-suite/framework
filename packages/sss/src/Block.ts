/* eslint-disable no-dupe-class-members, lines-between-class-members */

import { objectLoop } from '@aesthetic/utils';
import { Value } from './types';

export default class Block<T extends object = object> {
  readonly nested: { [selector: string]: Block<T> } = {};

  readonly properties: { [prop: string]: Value } = {};

  readonly selector: string;

  readonly variables: { [name: string]: Value } = {};

  constructor(selector: string) {
    this.selector = selector;
  }

  addNested(block: Block<T>, merge: boolean = true): this {
    if (merge && this.nested[block.selector]) {
      this.nested[block.selector].merge(block);
    } else {
      this.nested[block.selector] = block;
    }

    return this;
  }

  addProperty<K extends keyof T>(key: K, value: T[K]): this;
  addProperty(key: string, value: Value): this;
  addProperty(key: string, value: unknown): this {
    this.properties[key] = value as string;

    return this;
  }

  addProperties(properties: Partial<T> | { [key: string]: Value }): this;
  addProperties(properties: { [key: string]: unknown }): this {
    Object.assign(this.properties, properties);

    return this;
  }

  addVariable(key: string, value: Value): this {
    this.variables[key] = value;

    return this;
  }

  clone(selector: string): Block<T> {
    return new Block<T>(selector).merge(this);
  }

  merge(block: Block<T>): this {
    this.addProperties(block.properties);

    objectLoop(block.nested, (nested) => {
      this.addNested(nested);
    });

    return this;
  }

  toObject<O extends object = T>(): O {
    const object: { [key: string]: unknown } = {
      ...this.variables,
      ...this.properties,
    };

    objectLoop(this.nested, (block, selector) => {
      object[selector] = block.toObject();
    });

    return object as O;
  }
}
