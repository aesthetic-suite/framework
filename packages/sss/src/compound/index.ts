import animation from './animation';
import animationName from './animationName';
import fontFamily from './fontFamily';
import transition from './transition';

export default { animation, animationName, fontFamily, transition };
