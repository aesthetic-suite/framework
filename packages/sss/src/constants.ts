import { PagePseudos } from './types';

export const COMPOUND_PROPERTIES: Set<string> = new Set(['animationName', 'fontFamily']);

export const EXPANDED_PROPERTIES: Set<string> = new Set([
  'animation',
  'background',
  'border',
  'borderBottom',
  'borderLeft',
  'borderRight',
  'borderTop',
  'columnRule',
  'flex',
  'font',
  'listStyle',
  'margin',
  'offset',
  'outline',
  'padding',
  'textDecoration',
  'transition',
]);

export const PAGE_PSEUDOS: PagePseudos[] = [':blank', ':first', ':left', ':right'];
