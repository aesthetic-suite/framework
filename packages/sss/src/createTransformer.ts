import { isObject } from '@aesthetic/utils';
import { Transformer, TransformerHandler, Properties, TransformerUtils, Length } from './types';

export default function createTransformer<T>(
  property: keyof Properties,
  transformer?: Transformer<T>,
  unique?: boolean,
): TransformerHandler<T> {
  return (prop, wrap, customTransformer) => {
    const utils: TransformerUtils = {
      join(...values) {
        return values.filter(Boolean).join(' ');
      },
      separate(...values) {
        return values.filter(Boolean).join(' / ');
      },
      wrap(value) {
        return wrap(property, value);
      },
    };

    const handler = (item: T) => {
      if (isObject(item)) {
        if (customTransformer) {
          return customTransformer(item, utils);
        }

        if (transformer) {
          return transformer(prop, utils);
        }
      }

      return (item as unknown) as Length;
    };

    if (Array.isArray(prop)) {
      const result = prop.map(handler);

      return (unique ? Array.from(new Set(result)) : result).join(', ');
    }

    return handler(prop);
  };
}
