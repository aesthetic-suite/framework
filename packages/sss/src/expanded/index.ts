import animation from './animation';
import background from './background';
import border from './border';
import columnRule from './columnRule';
import flex from './flex';
import font from './font';
import listStyle from './listStyle';
import margin from './margin';
import offset from './offset';
import outline from './outline';
import padding from './padding';
import textDecoration from './textDecoration';
import transition from './transition';

export default {
  animation,
  background,
  border,
  columnRule,
  flex,
  font,
  listStyle,
  margin,
  offset,
  outline,
  padding,
  textDecoration,
  transition,
};
