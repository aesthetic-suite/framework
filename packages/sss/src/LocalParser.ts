import { isObject, objectLoop } from '@aesthetic/utils';
import Parser, { CommonEvents } from './Parser';
import Block from './Block';
import { LocalStyleSheet, ClassNameListener, RulesetListener, ParserOptions } from './types';

export const CLASS_NAME = /^[a-z]{1}[a-z0-9-_]+$/iu;

export interface LocalEvents<T extends object> extends CommonEvents<T> {
  onClass?: ClassNameListener;
  onRuleset?: RulesetListener<T>;
}

export default class LocalParser<T extends object> extends Parser<T, LocalEvents<T>> {
  parse(styleSheet: LocalStyleSheet, options: ParserOptions = {}) {
    Object.assign(this.options, options);

    objectLoop(styleSheet, (declaration, selector) => {
      // At-rule
      if (selector.charAt(0) === '@') {
        if (__DEV__) {
          throw new SyntaxError(
            `At-rules may not be defined at the root of a local block, found "${selector}".`,
          );
        }

        // Class name
      } else if (typeof declaration === 'string' && declaration.match(CLASS_NAME)) {
        this.emit('class', selector, declaration);

        // Declaration
      } else if (isObject(declaration)) {
        this.emit('ruleset', selector, this.parseLocalBlock(new Block(selector), declaration));

        // Unknown
      } else if (__DEV__) {
        throw new Error(
          `Invalid declaration for "${selector}". Must be an object (style declaration) or string (class name).`,
        );
      }
    });
  }
}
