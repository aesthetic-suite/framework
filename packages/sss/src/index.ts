/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Block from './Block';

export { Block };
