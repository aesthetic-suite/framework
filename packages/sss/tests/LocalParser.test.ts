import LocalParser from '../src/LocalParser';
import Block from '../src/Block';
import { Properties } from '../src/types';
import { createBlock } from './helpers';
import {
  SYNTAX_FALLBACKS,
  SYNTAX_PROPERTIES,
  SYNTAX_SELECTOR_ATTRIBUTES,
  SYNTAX_SELECTOR_PSEUDOS,
  SYNTAX_SELECTORS_SPECIFICITY,
  SYNTAX_SELECTORS_COMBINATORS,
  SYNTAX_SELECTORS_MULTIPLE,
  SYNTAX_SUPPORTS,
  SYNTAX_MEDIA,
  SYNTAX_MEDIA_NESTED,
} from './__mocks__/local';

describe('LocalParser', () => {
  let parser: LocalParser<Properties>;
  let spy: jest.Mock;

  beforeEach(() => {
    parser = new LocalParser();
    spy = jest.fn();
  });

  it('errors for an at-rule', () => {
    expect(() => {
      parser.parse({
        '@rule': {},
      });
    }).toThrow('At-rules may not be defined at the root of a local block, found "@rule".');
  });

  it('errors for invalid value type', () => {
    expect(() => {
      parser.parse({
        // @ts-ignore Allow invalid type
        el: 123,
      });
    }).toThrow(
      'Invalid declaration for "el". Must be an object (style declaration) or string (class name).',
    );
  });

  describe('class names', () => {
    beforeEach(() => {
      parser.on('class', spy);
    });

    it('emits for each class name', () => {
      parser.parse({
        foo: 'foo',
        bar: {},
        baz: 'baz',
      });

      expect(spy).toHaveBeenCalledTimes(2);
      expect(spy).toHaveBeenCalledWith('foo', 'foo');
      expect(spy).toHaveBeenCalledWith('baz', 'baz');
    });
  });

  describe('properties', () => {
    beforeEach(() => {
      parser.on('block:property', spy);
    });

    it('emits each property and value', () => {
      parser.parse({
        props: SYNTAX_PROPERTIES,
      });

      expect(spy).toHaveBeenCalledTimes(4);
      expect(spy).toHaveBeenCalledWith(expect.any(Block), 'color', 'black');
      expect(spy).toHaveBeenCalledWith(expect.any(Block), 'display', 'inline');
      expect(spy).toHaveBeenCalledWith(expect.any(Block), 'marginRight', 10);
      expect(spy).toHaveBeenCalledWith(expect.any(Block), 'padding', 0);
    });

    it('doesnt emit for undefined values', () => {
      parser.parse({
        props: {
          color: undefined,
        },
      });

      expect(spy).toHaveBeenCalledTimes(0);
    });
  });

  describe('attributes', () => {
    beforeEach(() => {
      parser.on('block:attribute', spy);
    });

    it('emits each attribute with value and params', () => {
      parser.parse({
        attrs: SYNTAX_SELECTOR_ATTRIBUTES,
      });

      expect(spy).toHaveBeenCalledTimes(2);

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        '[disabled]',
        createBlock('[disabled]', SYNTAX_SELECTOR_ATTRIBUTES['[disabled]']),
        { specificity: 0 },
      );

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        '[href]',
        createBlock('[href]', SYNTAX_SELECTOR_ATTRIBUTES['[href]']),
        { specificity: 0 },
      );
    });
  });

  describe('pseudos', () => {
    beforeEach(() => {
      parser.on('block:pseudo', spy);
    });

    it('emits each pseudo (class and element) with value and params', () => {
      parser.parse({
        attrs: SYNTAX_SELECTOR_PSEUDOS,
      });

      expect(spy).toHaveBeenCalledTimes(2);

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        ':hover',
        createBlock(':hover', SYNTAX_SELECTOR_PSEUDOS[':hover']),
        { specificity: 0 },
      );

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        '::before',
        createBlock('::before', SYNTAX_SELECTOR_PSEUDOS['::before']),
        { specificity: 0 },
      );
    });
  });

  describe('selectors', () => {
    beforeEach(() => {
      parser.on('block:selector', spy);
    });

    it('errors if selector is not an object', () => {
      expect(() => {
        parser.parse({
          selector: {
            // @ts-ignore Allow invalid type
            ':hover': 123,
          },
        });
      }).toThrow('":hover" must be a declaration object of CSS properties.');
    });

    it('errors if a comma separated list is passed', () => {
      expect(() => {
        parser.parse({
          selector: {
            // @ts-ignore Allow invalid type
            ':hover, :focus': {},
          },
        });
      }).toThrow('Advanced selector ":hover, :focus" must be nested within a @selectors block.');
    });
  });

  describe('@fallbacks', () => {
    beforeEach(() => {
      parser.on('block:fallback', spy);
    });

    it('errors if fallbacks are not an object', () => {
      expect(() => {
        parser.parse({
          fb: {
            // @ts-ignore Allow invalid type
            '@fallbacks': 123,
          },
        });
      }).toThrow('"@fallback" must be a declaration object of CSS properties.');
    });

    it('does not emit if no fallbacks', () => {
      parser.parse({
        fb: {},
      });

      expect(spy).not.toHaveBeenCalled();
    });

    it('emits for each fallback declaration', () => {
      parser.parse({
        fb: SYNTAX_FALLBACKS,
      });

      expect(spy).toHaveBeenCalledTimes(3);
      expect(spy).toHaveBeenCalledWith(expect.any(Block), 'background', ['red']);
      expect(spy).toHaveBeenCalledWith(expect.any(Block), 'display', ['block', 'inline-block']);
      expect(spy).toHaveBeenCalledWith(expect.any(Block), 'color', ['blue']);
    });
  });

  describe('@media', () => {
    beforeEach(() => {
      parser.on('block:media', spy);
    });

    it('errors if media is not an object', () => {
      expect(() => {
        parser.parse({
          fb: {
            // @ts-ignore Allow invalid type
            '@media': 123,
          },
        });
      }).toThrow('@media must be a mapping of CSS declarations.');
    });

    it('does not emit if no media', () => {
      parser.parse({
        media: {},
      });

      expect(spy).not.toHaveBeenCalled();
    });

    it('emits for each condition', () => {
      parser.parse({
        media: SYNTAX_MEDIA,
      });

      expect(spy).toHaveBeenCalledTimes(2);

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        '(min-width: 300px)',
        createBlock('@media (min-width: 300px)', {
          color: 'blue',
          paddingLeft: 15,
        }),
      );

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        '(max-width: 1000px)',
        createBlock('@media (max-width: 1000px)', {
          color: 'green',
          paddingLeft: 20,
        }),
      );
    });

    it('supports nested media conditions', () => {
      parser.parse({
        media: SYNTAX_MEDIA_NESTED,
      });

      expect(spy).toHaveBeenCalledTimes(2);

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        '(min-width: 300px)',
        createBlock('@media (min-width: 300px)', {
          color: 'blue',
          '@media (max-width: 1000px)': {
            color: 'green',
          },
        }),
      );

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        '(max-width: 1000px)',
        createBlock('@media (max-width: 1000px)', {
          color: 'green',
        }),
      );
    });
  });

  describe('@selectors', () => {
    beforeEach(() => {
      parser.on('block:selector', spy);
    });

    it('errors if selectors are not an object', () => {
      expect(() => {
        parser.parse({
          fb: {
            // @ts-ignore Allow invalid type
            '@selectors': 123,
          },
        });
      }).toThrow('@selectors must be a mapping of CSS declarations.');
    });

    it('does not emit if no selectors', () => {
      parser.parse({
        selectors: {},
      });

      expect(spy).not.toHaveBeenCalled();
    });

    it('emits selector with defined specificity', () => {
      const attrSpy = jest.fn();
      const pseudoSpy = jest.fn();

      parser.on('block:attribute', attrSpy);
      parser.on('block:pseudo', pseudoSpy);
      parser.parse({
        selectors: SYNTAX_SELECTORS_SPECIFICITY,
      });

      expect(pseudoSpy).toHaveBeenCalledTimes(2);

      expect(pseudoSpy).toHaveBeenCalledWith(
        expect.any(Block),
        ':hover',
        createBlock(':hover', { position: 'static' }),
        { specificity: 2 },
      );

      expect(pseudoSpy).toHaveBeenCalledWith(
        expect.any(Block),
        ':active',
        createBlock(':active', { position: 'absolute' }),
        { specificity: 1 },
      );

      expect(attrSpy).toHaveBeenCalledTimes(1);

      expect(attrSpy).toHaveBeenCalledWith(
        expect.any(Block),
        '[hidden]',
        createBlock('[hidden]', { position: 'relative' }),
        { specificity: 3 },
      );
    });

    it('emits universal and combinator selectors', () => {
      parser.parse({
        selectors: SYNTAX_SELECTORS_COMBINATORS,
      });

      expect(spy).toHaveBeenCalledTimes(4);

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        '> li',
        createBlock('> li', { listStyle: 'bullet' }),
        { specificity: 0 },
      );

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        '+ div',
        createBlock('+ div', { display: 'none' }),
        { specificity: 0 },
      );

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        '~ span',
        createBlock('~ span', { color: 'black' }),
        { specificity: 0 },
      );

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        '*',
        createBlock('*', { backgroundColor: 'inherit' }),
        { specificity: 0 },
      );
    });

    it('emits each selector in a comma separated list', () => {
      const attrSpy = jest.fn();
      const pseudoSpy = jest.fn();

      parser.on('block:attribute', attrSpy);
      parser.on('block:pseudo', pseudoSpy);
      parser.parse({
        selectors: SYNTAX_SELECTORS_MULTIPLE,
      });

      expect(pseudoSpy).toHaveBeenCalledWith(
        expect.any(Block),
        ':disabled',
        createBlock(':disabled', { cursor: 'default' }),
        { specificity: 0 },
      );

      expect(attrSpy).toHaveBeenCalledWith(
        expect.any(Block),
        '[disabled]',
        createBlock('[disabled]', { cursor: 'default' }),
        { specificity: 2 },
      );

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        '> span',
        createBlock('> span', { cursor: 'default' }),
        { specificity: 0 },
      );
    });
  });

  describe('@supports', () => {
    beforeEach(() => {
      parser.on('block:supports', spy);
    });

    it('errors if supports are not an object', () => {
      expect(() => {
        parser.parse({
          fb: {
            // @ts-ignore Allow invalid type
            '@supports': 123,
          },
        });
      }).toThrow('@supports must be a mapping of CSS declarations.');
    });

    it('does not emit if no supports', () => {
      parser.parse({
        supports: {},
      });

      expect(spy).not.toHaveBeenCalled();
    });

    it('emits for each condition', () => {
      parser.parse({
        supports: SYNTAX_SUPPORTS,
      });

      expect(spy).toHaveBeenCalledTimes(2);

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        '(display: flex)',
        createBlock('@supports (display: flex)', { display: 'flex' }),
      );

      expect(spy).toHaveBeenCalledWith(
        expect.any(Block),
        'not (display: flex)',
        createBlock('@supports not (display: flex)', { float: 'left' }),
      );
    });
  });
});
