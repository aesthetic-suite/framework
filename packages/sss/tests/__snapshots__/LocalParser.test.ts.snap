// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocalParser @variants errors for invalid compound name 1`] = `"Invalid variant \\"type:value + broken\\". Type and enumeration must be separated with a \\":\\", and each part may only contain a-z, 0-9, -, _."`;

exports[`LocalParser @variants errors if contains a space 1`] = `"Invalid variant \\"type:va lue\\". Type and enumeration must be separated with a \\":\\", and each part may only contain a-z, 0-9, -, _."`;

exports[`LocalParser @variants errors if enum is empty 1`] = `"Invalid variant \\"type:\\". Type and enumeration must be separated with a \\":\\", and each part may only contain a-z, 0-9, -, _."`;

exports[`LocalParser @variants errors if missing a colon 1`] = `"Invalid variant \\"foo\\". Type and enumeration must be separated with a \\":\\", and each part may only contain a-z, 0-9, -, _."`;

exports[`LocalParser @variants errors if starts with a number 1`] = `"Invalid variant \\"9a:value\\". Type and enumeration must be separated with a \\":\\", and each part may only contain a-z, 0-9, -, _."`;
